File: Project/Generic_DS_Lib/Queue.c
// Queue

#include <stdio.h>
#include <stdlib.h>

/*
1.Enqueue()
2.Dequeue()
3.Count()
4.Display()
*/

struct node
{
    int data;
    struct node * next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void Display(PNODE Head)
{
    while(Head != NULL)
    {
        printf("|%d|->",Head->data);
        Head = Head->next;
    }
    printf("NULL\n");
}

int Count(PNODE Head)
{
    int iCnt = 0;

    while(Head != NULL)
    {
        iCnt++;
        Head = Head->next;
    }
    return iCnt;
}

void Enqueue(PPNODE Head, int No)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->data = No;
    newn->next = NULL;

    if(*Head == NULL)
    {
        *Head = newn;
    }
    else
    {
        temp = *Head;

        while(temp->next != NULL)
        {
            temp = temp->next;
        }
        temp->next = newn;
    }
}

void Dequeue(PPNODE Head)
{
    int no = 0;
    PNODE temp = NULL;

    if(*Head == NULL)
    {
        printf("Queue is empty\n");
        return;
    }
    else if((*Head)->next == NULL)
    {
        no = (*Head)->data;
        free(*Head);
        *Head = NULL;
        printf("Removed element is : %d\n",no);
    }
    else
    {
        temp = *Head;
        no = (*Head)->data;
        *Head = (*Head)->next;
        free(temp);
        printf("Removed element is : %d\n",no);
    }
}

int main()
{
    int Ret = 0;
    PNODE First = NULL;

    Enqueue(&First,11);
    Enqueue(&First,21);
    Enqueue(&First,51);
    
    Display(First);
    Ret = Count(First);
    printf("Number of elements in queue are : %d\n",Ret);

    Dequeue(&First);
    Dequeue(&First);
    Display(First);
    Ret = Count(First);
    printf("Number of elements in queue are : %d\n",Ret);

    Dequeue(&First);
    Dequeue(&First);
    Display(First);
    Ret = Count(First);
    printf("Number of elements in queue are : %d\n",Ret);

    return 0;
}

File: Project/Generic_DS_Lib/Queue.java
import java.lang.*;
import java.util.*;

class node
{
    public int data;
    public node next;       // struct node * next;
}

class QueueDS
{
    public node Head;
    public int Count;

    public QueueDS()
    {
        Head = null;
        Count = 0;
    }

    public void Enqueue(int no)
    {
        node newn = new node();

        newn.data = no;
        newn.next = null;

        if(Head == null)
        {
            Head = newn;
        }
        else
        {
            node temp = Head;

            while(temp.next != null)
            {
                temp = temp.next;
            }
            temp.next = newn;
        }
        Count++;
    }

    public void Dequeue()
    {
        int no;
        if(Count == 0)
        {
            System.out.println("Queue is empty");
            return;
        }
        else if(Count == 1)
        {
            no = Head.data;
            Head = null;
        }
        else
        {
            no = Head.data;
            node temp = Head;
            Head = Head.next;
            temp = null;
        }
        System.out.println("Removed element is : "+no);
        Count--;
    }

    public void Display()
    {
        node temp = Head;

        System.out.print("Elements of queue are : ");
        while(temp != null)
        {
            System.out.print("|"+temp.data+"|-");
            temp = temp.next;
        }
        System.out.println("NULL");
    }

    public int CountNode()
    {
        return Count;
    }   
}

class Queue
{
    public static void main(String arg[])
    {
        QueueDS obj = new QueueDS();

        obj.Enqueue(11);
        obj.Enqueue(21);
        obj.Enqueue(51);

        obj.Display();
        System.out.println("Number of elements in queue are : "+obj.CountNode());

        obj.Dequeue();
        obj.Dequeue();

        obj.Display();
        System.out.println("Number of elements in queue are : "+obj.CountNode());

        obj.Dequeue();
        obj.Dequeue();

        obj.Display();
        System.out.println("Number of elements in queue are : "+obj.CountNode());
    }
}